addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/grxprg/SocketSession.java

```diff
@@ -0,0 +1,385 @@
+     * {@link #_client}
+     * will be closed and set to null. The {@link #_dispatcher} and {@link #_responseReader} will be stopped, the
+     * {@link #_callback} will be nulled and the {@link #_responses} will be cleared.
+     *
+     * @throws java.io.IOException if an exception occurs during the disconnect attempt
+     */
+    public void disconnect() throws IOException {
+        if (isConnected()) {
+            _logger.debug("Disconnecting from {}:{}", _host, _port);
+
+            _dispatcher.stopRunning();
+            _responseReader.stopRunning();
+
+            _writer.close();
+            _writer = null;
+
+            _reader.close();
+            _reader = null;
+
+            _client.close();
+            _client = null;
+
+            _callback.set(null);
+            _responses.clear();
+        }
+    }
+
+    /**
+     * Returns true if we are connected ({@link #_client} is not null and is connected)
+     *
+     * @return true if connected, false otherwise
+     */
+    public boolean isConnected() {
+        return _client != null && _client.isConnected();
+    }
+
+    /**
+     * Sends the specified command to the underlying socket
+     *
+     * @param command a non-null, non-empty command
+     * @throws java.io.IOException an exception that occurred while sending
+     */
+    public synchronized void sendCommand(String command) throws IOException {
+        if (command == null) {
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        if (!isConnected()) {
+            throw new IOException("Cannot send message - disconnected");
+        }
+
+        _logger.debug("Sending Command: '{}'", command);
+        _writer.println(command + "\n"); // as pre spec - each command must have a newline
+        _writer.flush();
+
+    }
+
+    /**
+     * This is the runnable that will read from the socket and add messages to the responses queue (to be processed by
+     * the dispatcher)
+     *
+     * @author Tim Roberts
+     *
+     */
+    private class ResponseReader implements Runnable {
+
+        /**
+         * Whether the reader is currently running
+         */
+        private final AtomicBoolean _isRunning = new AtomicBoolean(false);
+
+        /**
+         * Locking to allow proper shutdown of the reader
+         */
+        private final Lock _lock = new ReentrantLock();
+        private final Condition _running = _lock.newCondition();
+
+        /**
+         * Stops the reader. Will wait 5 seconds for the runnable to stop (should stop within 1 second based on the
+         * setSOTimeout)
+         */
+        public void stopRunning() {
+            _lock.lock();
+            try {
+                if (_isRunning.getAndSet(false)) {
+                    if (!_running.await(5, TimeUnit.SECONDS)) {
+                        _logger.warn("Waited too long for dispatcher to finish");
+                    }
```

Comment:
Should this be closing the socket or something?  It's not clear to me how this will stop the thread unless some data comes down the wire and brings the thread out of the blocking read().